libs/algo/odc/algo/_tiff.py

```diff
@@ -0,0 +1,518 @@
+            block = item
+        elif ndim == 3:
+            if info.axis == 0:
+                # Y, X, B
+                bands = np_slice_to_idx(key[2], info.count)
+                if item.ndim == 2:
+                    block = np.expand_dims(item, axis=0)
+                else:
+                    # rio expects band to be the first dimension
+                    block = item.transpose([2, 0, 1])
+            else:
+                # B, Y, X
+                bands = np_slice_to_idx(key[0], info.count)
+                if item.ndim == 2:
+                    block = np.expand_dims(item, axis=0)
+                else:
+                    block = item
+
+            # rio wants 1 based indexing
+            bands = tuple(i + 1 for i in bands)
+        else:
+            raise ValueError("Only accept 2 and 3 dimensional data")
+
+        win = Window.from_slices(*yx_key, height=info.height, width=info.width)
+        if self._lock:
+            with self._lock:
+                self._out.write(block, indexes=bands, window=win)
+        else:
+            self._out.write(block, indexes=bands, window=win)
+
+
+class COGSink:
+    def __init__(
+        self,
+        info: GeoRasterInfo,
+        dst: str,
+        blocksize: Optional[int] = None,
+        ovr_blocksize: Optional[int] = None,
+        bigtiff: Union[bool, str] = "auto",
+        lock: bool = True,
+        temp_folder: Optional[str] = None,
+        overview_resampling: str = "average",
+        rio_opts_first_pass: Optional[Dict[str, Any]] = None,
+        use_final_blocksizes: bool = False,
+        **extra_rio_opts,
+    ):
+        if blocksize is None:
+            blocksize = 512
+
+        if ovr_blocksize is None:
+            ovr_blocksize = blocksize
+
+        if bigtiff == "auto":
+            # do bigtiff if raw raster is larger than 4GB
+            bigtiff = info.raster_size() > (1 << 32)
+
+        opts = dict(
+            driver="GTiff",
+            bigtiff=bigtiff,
+            tiled=True,
+            blockxsize=_adjust_blocksize(blocksize, info.width),
+            blockysize=_adjust_blocksize(blocksize, info.height),
+            compress="DEFLATE",
+            zlevel=6,
+            predictor=2,
+            num_threads="ALL_CPUS",
+        )
+        opts.update(extra_rio_opts)
+
+        if rio_opts_first_pass is None:
+            rio_opts_first_pass = dict(
+                compress="zstd",
+                zstd_level=1,
+                predictor=1,
+                num_threads="ALL_CPUS",
+                sparse_ok=True,
+                interleave=opts.get("interleave", "pixel"),
+            )
+
+        layers = []
+        temp = str(uuid4())
+        t_dir = ""
+        if temp_folder:
+            t_name = temp
+        else:
+            t_dir, t_name = temp[:8], temp[9:]
+
+        ext = ".tif"
+        ii = info
+        bsz = 2048
+        for idx in range(7 + 1):
+            if temp_folder:
```

Comment:
I think this can be pulled out of the loop

Reply:
agree that this can be structured better, but not sure creating factory method is worth it here.

It's creating this:
```
  /path/file.tif
  /path/file.tif.ovr
  /path/file/tif.ovr.ovr
....
```
and so on. Some times it's just a path other times it's `MemoryFile`

Reply:
oh wait it's creating a new `MemoryFile` with the same `dirname` and `filename` each loop - are these distinct objects even though they have the same name and location?